circuits/cpp/src/aztec3/circuits/kernel/private/private_kernel_circuit.cpp

```diff
@@ -261,6 +263,41 @@ void validate_inputs(PrivateInputs<CT> const& private_inputs)
     CT::fr const start_public_call_stack_length = array_length<Composer>(start.public_call_stack);
     CT::fr const start_new_l2_to_l1_msgs_length = array_length<Composer>(start.new_l2_to_l1_msgs);
 
+    {
+        // Verify public key hash matches ethereum address of the sender
+        // TODO(Suyash): We want to perform this only for the base case, so move to "initial" PKC.
+        const CT::address sender_address = private_inputs.signed_tx_request.tx_request.from;
+        const CT::secp256k1_point sender_public_key = private_inputs.signed_tx_request.signing_key;
+
+        // If the sender public key is P = (x, y), Ethereum address is computed as:
+        // E = keccak([x || y]).slice(12, 32)
+        CT::byte_array sender_public_key_bytes(sender_public_key.get_context());
+        sender_public_key_bytes.write(sender_public_key.x.to_byte_array());
+        sender_public_key_bytes.write(sender_public_key.y.to_byte_array());
+        const CT::byte_array sender_public_key_hash = stdlib::keccak<Composer>::hash(sender_public_key_bytes);
+        const CT::byte_array sender_address_bytes = CT::byte_array(sender_address.to_field());
+
+        // Check if the sender address matches the keccak hash of the public key
+        // Specifically, first 12 bytes must be 0, remaining 20 bytes must match.
+        for (size_t i = 0; i < 12; i++) {
+            sender_address_bytes[i].assert_is_zero(format("sender address at index ", i, " is non-zero"));
+        }
+        for (size_t i = 12; i < 32; i++) {
+            sender_address_bytes[i].assert_equal(
+                sender_public_key_hash[i], format("hash of public key does not match the sender address at index ", i));
+        }
+    }
+
+    // Verify signature against the first function being called (subsequent function calls do not
+    // need to be signed over by the sender, the sender only signs the inputs to the very first function)
+    const CT::byte_array message = private_inputs.signed_tx_request.tx_request.hash();
+    auto sig_verification_result = stdlib::ecdsa::verify_signature<Composer,
+                                                                   stdlib::secp256k1<Composer>,
+                                                                   stdlib::secp256k1<Composer>::fq_ct,
+                                                                   stdlib::secp256k1<Composer>::bigfr_ct,
+                                                                   stdlib::secp256k1<Composer>::g1_bigfr_ct>(
+        message, private_inputs.signed_tx_request.signing_key, private_inputs.signed_tx_request.signature);
+
```

Comment:
Good work! Tbh our private kernel might be far behind enough that we might want to consider not maintaining it until we are ready to line-by-line compare with native and update it.

But for the ECDSA work, you needed to test the circuit-implementation so it is good to have this here :+1:

Reply:
Yeah exactly.